File: javascript/beer-song/beer.js
var refill = 99
  , template = "$current of beer on the wall, $current of beer.\n" +
               "$action, $remaining of beer on the wall.\n";

module.exports = {
  sing: function(from, to) {
    var verses = [];
    for (var n = from; n >= (to || 0); n--)
      verses.push(this.verse(n));
    return verses.join("\n");
  },

  verse: function(n) {
    return ucfirst(template
      .replace(/\$current/g,   formatQuantity(n))
      .replace(/\$action/g,    formatAction(n))
      .replace(/\$remaining/g, formatQuantity(n === 0 ? refill : n - 1)));
  }
};

function formatAction(n) {
  if (n === 0)
    return "Go to the store and buy some more";
  return ["Take", n === 1 ? "it" : "one", "down and pass it around"].join(" ");
}

function formatQuantity(n) {
  if (n === 0)
    return "no more bottles";
  return n + " bottle" + (n > 1 ? "s" : "");
}

function ucfirst(str) {
  return str.charAt(0).toUpperCase() + str.slice(1);
}

File: elixir/beer-song/beer.exs
defmodule Beer do
  def verse(0),    do: verse(0, 99)
  def verse(n),    do: verse(n, n - 1)
  def verse(n, r), do: render(n, r)

  def sing(from),     do: sing(from, 0)
  def sing(from, to), do: assemble(from, to)

  defp format_bottles(0), do: "no more bottles of beer"
  defp format_bottles(1), do: "1 bottle of beer"
  defp format_bottles(n), do: "#{n} bottles of beer"

  defp format_action(0), do: "Go to the store and buy some more"
  defp format_action(1), do: "Take it down and pass it around"
  defp format_action(n), do: "Take one down and pass it around"

  defp render(n, remaining) do
    bottles = format_bottles(n)
    action = format_action(n)
    remaining = format_bottles(remaining)
    String.capitalize("#{bottles} on the wall, #{bottles}.\n")
                   <> "#{action}, #{remaining} on the wall.\n"
  end

  defp assemble(from, to) do
    Enum.map_join((from..to), "\n", verse &1) <> "\n"
  end
end
